### ISSUE TITLE
Duplicate types for "Keys": Code generation bug, not able to compile

### ISSUE THREAD

**Issue by Miladiir** [prisma/prisma#10490]
### Bug description

Code generation produces duplicate types for a reasonable prisma schema. One of the types is a typescript helper type, generated or included by prisma, the other is the obviously expected type for typescript typings in the prisma client.

Typescript Error:
<details>

```
node_modules/.prisma/client/index.d.ts:46171:11 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46171         ? Keys
                ~~~~

node_modules/.prisma/client/index.d.ts:46176:7 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46176     ? Keys
            ~~~~

node_modules/.prisma/client/index.d.ts:46179:51 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46179     [P in TrueKeys<S['select']>]: P extends keyof Keys ?Keys [P]
                                                        ~~~~

node_modules/.prisma/client/index.d.ts:46179:57 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46179     [P in TrueKeys<S['select']>]: P extends keyof Keys ?Keys [P]
                                                              ~~~~

node_modules/.prisma/client/index.d.ts:46183:7 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46183     : Keys
            ~~~~

node_modules/.prisma/client/index.d.ts:46184:5 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46184   : Keys
          ~~~~

node_modules/.prisma/client/index.d.ts:46207:132 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46207     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>

                                  ~~~~

node_modules/.prisma/client/index.d.ts:46207:214 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46207     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>


             ~~~~

node_modules/.prisma/client/index.d.ts:46224:131 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46224     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>

                                 ~~~~

node_modules/.prisma/client/index.d.ts:46224:213 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46224     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>


            ~~~~

node_modules/.prisma/client/index.d.ts:46244:43 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46244     ): CheckSelect<T, PrismaPromise<Array<Keys>>, PrismaPromise<Array<KeysGetPayload<T>>>>
                                                ~~~~

node_modules/.prisma/client/index.d.ts:46260:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46260     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                               ~~~~

node_modules/.prisma/client/index.d.ts:46292:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46292     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                               ~~~~

node_modules/.prisma/client/index.d.ts:46311:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46311     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                               ~~~~

node_modules/.prisma/client/index.d.ts:46369:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

46369     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                               ~~~~


Found 15 errors.
```
</details>

The generated file is attached. CTRL + F for 'type Keys'
[index.d.zip](https://github.com/prisma/prisma/files/7619551/index.d.zip)

### How to reproduce

1. Empty Project
2. Install typescript, prisma and @prisma/client
3. Copy and paste the three files below
4. `npx prisma db push`
5. `npx prisma generate`
6. `npx tsc`

Observed Error: The helper type `Keys` conflicts with the generated type of the database table.

### Expected behavior

The generated code works and does not include helper types with names that clash with possible table names. Maybe generate the types with a prefix.

### Prisma information

schema.prisma
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model Keys {
  KeyType  String
  KeyCode  Int
  Language String
  KeyText  String?

  @@id([KeyType, KeyCode, Language])
}

```

index.ts
```typescript
import { PrismaClient } from "@prisma/client";
const prisma = new PrismaClient();

prisma.keys.findMany().then((key) => console.log(key));
```

tsconfig.json
```json
{
  "compilerOptions": {
    "target": "es2016" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
    "module": "commonjs" /* Specify what module code is generated. */,
    "esModuleInterop": true /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables `allowSyntheticDefaultImports` for type compatibility. */,
    "forceConsistentCasingInFileNames": true /* Ensure that casing is correct in imports. */,
    "strict": true /* Enable all strict type-checking options. */
  }
}

```

### Environment & setup

- OS: Windows
- Database: SQLite (also tested with MSSQL, same error)
- Node.js version: v14.17.6

### Prisma Version

```
prisma                  : 3.5.0
@prisma/client          : 3.5.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine 78a5df6def6943431f4c022e1428dbc3e833cf8e (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli 78a5df6def6943431f4c022e1428dbc3e833cf8e (at node_modules\@prisma\engines\migration-engine-windows.exe)
Introspection Engine    : introspection-core 78a5df6def6943431f4c022e1428dbc3e833cf8e (at node_modules\@prisma\engines\introspection-engine-windows.exe)
Format Binary           : prisma-fmt 78a5df6def6943431f4c022e1428dbc3e833cf8e (at node_modules\@prisma\engines\prisma-fmt-windows.exe)
Default Engines Hash    : 78a5df6def6943431f4c022e1428dbc3e833cf8e
Studio                  : 0.439.0
```

**Comment by janpio** [comment]
I can confirm this with 3.6.0:

```
C:\Users\Jan\Documents\throwaway\10490  (10490@1.0.0)
λ npx tsc
node_modules/.prisma/client/index.d.ts:379:15 - error TS2580: Cannot find name 'Buffer'. Do you need to install type definitions for node? Try `npm i --save-dev @types/node`.

379   : T extends Buffer
                  ~~~~~~

node_modules/.prisma/client/index.d.ts:381:15 - error TS2583: Cannot find name 'BigInt'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2020' or later.

381   : T extends BigInt
                  ~~~~~~

node_modules/.prisma/client/index.d.ts:950:11 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

950         ? Keys
              ~~~~

node_modules/.prisma/client/index.d.ts:955:7 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

955     ? Keys
          ~~~~

node_modules/.prisma/client/index.d.ts:958:51 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

958     [P in TrueKeys<S['select']>]: P extends keyof Keys ?Keys [P]
                                                      ~~~~

node_modules/.prisma/client/index.d.ts:958:57 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

958     [P in TrueKeys<S['select']>]: P extends keyof Keys ?Keys [P]
                                                            ~~~~

node_modules/.prisma/client/index.d.ts:962:7 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

962     : Keys
          ~~~~

node_modules/.prisma/client/index.d.ts:963:5 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

963   : Keys
        ~~~~

node_modules/.prisma/client/index.d.ts:986:132 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

986     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>
                                                                                                                                       ~~~~

node_modules/.prisma/client/index.d.ts:986:214 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

986     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>

                                                              ~~~~

node_modules/.prisma/client/index.d.ts:1003:131 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1003     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>
                                                                                                                                       ~~~~

node_modules/.prisma/client/index.d.ts:1003:213 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1003     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Keys'> extends True ? CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>> : CheckSelect<T, Prisma__KeysClient<Keys | null >, Prisma__KeysClient<KeysGetPayload<T> | null >>

                                                              ~~~~

node_modules/.prisma/client/index.d.ts:1023:43 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1023     ): CheckSelect<T, PrismaPromise<Array<Keys>>, PrismaPromise<Array<KeysGetPayload<T>>>>
                                               ~~~~

node_modules/.prisma/client/index.d.ts:1039:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1039     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                              ~~~~

node_modules/.prisma/client/index.d.ts:1055:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1055     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                              ~~~~

node_modules/.prisma/client/index.d.ts:1074:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1074     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                              ~~~~

node_modules/.prisma/client/index.d.ts:1132:42 - error TS2314: Generic type 'Keys' requires 1 type argument(s).

1132     ): CheckSelect<T, Prisma__KeysClient<Keys>, Prisma__KeysClient<KeysGetPayload<T>>>
                                              ~~~~

node_modules/@prisma/client/runtime/index.d.ts:1:23 - error TS2688: Cannot find type definition file for 'node'.

1 /// <reference types="node" />
                        ~~~~

node_modules/@prisma/client/runtime/index.d.ts:3:25 - error TS2307: Cannot find module 'util' or its corresponding type declarations.

3 import { inspect } from 'util';
                          ~~~~~~

node_modules/@prisma/client/runtime/index.d.ts:1180:5 - error TS1165: A computed property name in an ambient context must refer to an expression whose type is a literal type or a 'unique symbol' type.

1180     [inspect.custom](): {
         ~~~~~~~~~~~~~~~~


Found 20 errors.
```

**Comment by jlenon7** [comment]
Same problem here:

```
node_modules/.prisma/client/index.d.ts:2490:119 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

2490     merges<T extends MergeFindManyArgs = {}>(args?: Subset<T, MergeFindManyArgs>): CheckSelect<T, PrismaPromise<Array<Merge>>, PrismaPromise<Array<MergeGetPayload<T>>>>;
                                                                                                                           ~~~~~

node_modules/.prisma/client/index.d.ts:4808:11 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4808         ? Merge
               ~~~~~

node_modules/.prisma/client/index.d.ts:4813:7 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4813     ? Merge  & {
           ~~~~~

node_modules/.prisma/client/index.d.ts:4820:51 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4820     [P in TrueKeys<S['select']>]: P extends keyof Merge ?Merge [P]
                                                       ~~~~~

node_modules/.prisma/client/index.d.ts:4820:58 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4820     [P in TrueKeys<S['select']>]: P extends keyof Merge ?Merge [P]
                                                              ~~~~~

node_modules/.prisma/client/index.d.ts:4825:7 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4825     : Merge
           ~~~~~

node_modules/.prisma/client/index.d.ts:4826:5 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4826   : Merge
         ~~~~~

node_modules/.prisma/client/index.d.ts:4849:134 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4849     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Merge'> extends True ? CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>> : CheckSelect<T, Prisma__MergeClient<Merge | null >, Prisma__MergeClient<MergeGetPayload<T> | null >>
                                                                                                                                          ~~~~~

node_modules/.prisma/client/index.d.ts:4849:220 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4849     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findUnique', 'Merge'> extends True ? CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>> : CheckSelect<T, Prisma__MergeClient<Merge | null >, Prisma__MergeClient<MergeGetPayload<T> | null >>
                                                                                                                                                                                                                                ~~~~~

node_modules/.prisma/client/index.d.ts:4866:133 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4866     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Merge'> extends True ? CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>> : CheckSelect<T, Prisma__MergeClient<Merge | null >, Prisma__MergeClient<MergeGetPayload<T> | null >>
                                                                                                                                         ~~~~~

node_modules/.prisma/client/index.d.ts:4866:219 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4866     ): HasReject<GlobalRejectSettings, LocalRejectSettings, 'findFirst', 'Merge'> extends True ? CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>> : CheckSelect<T, Prisma__MergeClient<Merge | null >, Prisma__MergeClient<MergeGetPayload<T> | null >>
                                                                                                                                                                                                                               ~~~~~

node_modules/.prisma/client/index.d.ts:4886:43 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4886     ): CheckSelect<T, PrismaPromise<Array<Merge>>, PrismaPromise<Array<MergeGetPayload<T>>>>
                                               ~~~~~

node_modules/.prisma/client/index.d.ts:4902:43 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4902     ): CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>>
                                               ~~~~~

node_modules/.prisma/client/index.d.ts:4934:43 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4934     ): CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>>
                                               ~~~~~

node_modules/.prisma/client/index.d.ts:4953:43 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

4953     ): CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>>
                                               ~~~~~

node_modules/.prisma/client/index.d.ts:5011:43 - error TS2314: Generic type 'Merge' requires 1 type argument(s).

5011     ): CheckSelect<T, Prisma__MergeClient<Merge>, Prisma__MergeClient<MergeGetPayload<T>>>
                                               ~~~~~

[9:33:51 AM] Found 16 errors. Watching for file changes.
```

**Comment by jlenon7** [comment]
For a quick fix I changed the name of the Table from `Merge` to `MergeRequest` to no conflict with the database types.

**Comment by Miladiir** [comment]
Probably simpler to disable type checking of `*.d.ts` files. This way, you don't have to design your db for prisma to work correctly. I still believe this should be fixed, preferably by renaming all helper types.

**Comment by SevInf** [comment]
This issue is now fixed and the fix will be published as a part of `5.1` release. If you'd like to verify fix earlier, you can use dev snapshot `5.1.0-dev.24`. We don't recommend using it in production, but it should be enough to verify the fix.
